#79: Creating Data Points For 3D Graph

Subzeero opened the pull request:
Closes #79 

# What was the issue
One of our core requirements for our project requires a plot for analyzing data. This PR includes one of the necessary components for this plot: the data points. 

# What was done and why
I created two React components: DataPoint and DataPointMenu. The DataPoint component will represent each point of data that will be plotted. The DataPointMenu appears when a DataPoint is selected and it display information about the DataPoint. The selected point is shared using a PointSelectionContext which stores the selected DataPoint id number in its state.

# How it was tested
I used the WebXR emulator to test the requested functionality.

Jesster2829 commented on my-webxr-app/src/App.tsx:
```diff
@@ -4,29 +4,35 @@ import '@react-three/fiber'
+            <ambientLight />
+            <pointLight position={[10, 10, 10]} />
+            <Controllers />
+            {/*<Button position={[0, 1.5, -1]} />*/}
+            {/*<RotatingBox position={[0.8, 1.5, -1]} />*/}
+            {/*<RotatingBox position={[-0.8, 1.5, -1]} />*/}
```
Can probably remove this commented out code

Jesster2829 commented on my-webxr-app/src/contexts/PointSelectionContext.tsx:
```diff
@@ -0,0 +1,38 @@
+import React, { createContext, useContext, useMemo, useState } from "react";
+
+interface PointSelectionContext {
+  selectedDataPoint: number | null;
+  setSelectedDataPoint: React.Dispatch<
+    React.SetStateAction<PointSelectionContext["selectedDataPoint"]>
+  >;
+}
+
+export const PointSelectionContext =
+  createContext<PointSelectionContext | null>(null);
+
+export const PointSelectionProvider = ({
+  children,
+}: React.PropsWithChildren) => {
+  const [selectedDataPoint, setSelectedDataPoint] =
+    useState<PointSelectionContext["selectedDataPoint"]>(null);
+  const value = useMemo(
+    () => ({ selectedDataPoint, setSelectedDataPoint }),
+    [selectedDataPoint],
+  );
+
+  return (
+    <PointSelectionContext.Provider value={value}>
+      {children}
+    </PointSelectionContext.Provider>
+  );
+};
+
+export const usePointSelectionContext = () => {
+  const context = useContext(PointSelectionContext);
+  if (!context) {
+    throw new Error(
+      "You must use this context within a PointSelectionProvider!",
+    );
+  }
+  return context;
+};
```
Add some documentation for what is happening here

Jesster2829 reviewed (changes requested):
Just add documentation otherwise good to go

Subzeero commented:
Addressed the review concerns. Also fixed the minor issue when two controllers hovered over the same DataPoint and the hover effect was prematurely removed. Simply switched from a Boolean state to a count of hovers.

LongQuanNguyen reviewed (commented):
The components folder is with 'C' capitalized. Should it be in lowercase?

matthew-buglass commented on my-webxr-app/src/components/DataPoint.tsx:
Needs unit tests

matthew-buglass commented on my-webxr-app/src/components/DataPointMenu.tsx:
needs unit tests

matthew-buglass commented on my-webxr-app/src/contexts/PointSelectionContext.tsx:
needs unit tests

matthew-buglass reviewed (changes requested):
We just need to add unit tests for the components

LongQuanNguyen commented on my-webxr-app/src/contexts/PointSelectionContext.tsx:
```diff
@@ -0,0 +1,60 @@
+/**
+ * Create the Context.
+ */
+export const PointSelectionContext =
+  createContext<PointSelectionContext | null>(null);
```
Yes, Comment resolve during pairing

LongQuanNguyen commented on my-webxr-app/src/contexts/PointSelectionContext.tsx:
```diff
@@ -0,0 +1,60 @@
+ * Call this function instead of useContext(PointSelectionContext).
+ */
+export const usePointSelectionContext = () => {
+  const context = useContext(PointSelectionContext);
```
The object is returned so it is guaranteed to be not null -- per discussion during paring

Jesster2829 reviewed (approved):
Pre approval before commit, have witnessed all changes in chat but will push after his assignment is done
